README.md
# ✨ Sparkla

> Compute sessions for executable documents: fast-to-start, isolated, reproducible. Pick any three.

> :warning: Well, that's the aim ;) Sparkla is currently in early development. It started as an experiment in using Amazon's Firecracker as a more secure (but just as fast-to-start) alternative to Docker containers. It is intended to supersede [`stencila/cloud`](https://github.com/stencila/cloud).

## Install

```bash
npm install -g @stencila/sparkla
```

Sparkla creates sessions in either Firecracker microVMs or Docker containers. So you'll need at least one of these installed.

### Firecracker

Firecracker is only available on Linux.
Download the latest Firecracker binary from https://github.com/firecracker-microvm/firecracker/releases:

```bash
curl -L -o firecracker https://github.com/firecracker-microvm/firecracker/releases/download/v0.18.0/firecracker-v0.18.0
chmod +x firecracker
```

Firecracker is built on top of KVM and needs read/write access to `/dev/kvm`. Log in to the host in one terminal and set up that access:

```bash
sudo setfacl -m u:${USER}:rw /dev/kvm
```

### Docker

If you want to use Docker-based sessions then you'll need to have `docker` installed.

## Usage

Run Sparkla using the command line interface e.g.

```bash
sparkla serve --port 9001 --cpuTotal 1
```

### Options

<!-- prettier-ignore-start -->
<!-- OPTIONS-BEGIN -->

| Name            | Description                                                                                                              | Type                           | Default              |
| --------------- | ------------------------------------------------------------------------------------------------------------------------ | ------------------------------ | -------------------- |
| debug           | Output debug level log entries?                                                                                          | boolean                        | false                |
| host            | The host address that the server should listen on.                                                                       | string                         | "127.0.0.1"          |
| port            | The port that the server should listen on.                                                                               | number                         | 9000                 |
| jwtSecret       | The JWT secret to use to sign and verify JWT tokens. If `null` then a random secret will be generated.                   | string, null                   |  null                |
| sessionType     | The class of sessions created.                                                                                           | "firecracker", "docker"        | "docker"             |
| cpuTotal        | The total number of CPUs that can be allocated to sessions. `null` = use the number of CPUs on the machine.              | number, null                   |  null                |
| memoryTotal     | The total number amount of memory (Gib) that can be allocated to sessions. `null` = use the total amount of memory on the machine. | number, null                   |  null                |
| expiryInterval  | Interval in seconds between checks for expired sessions.                                                                 | number                         | 15                   |
| durationWarning | Number of seconds to provide clients with a warning prior to reaching maximum session duration.                          | number                         | 600                  |
| timeoutWarning  | Number of seconds to provide clients with a warning prior to a reaching session timeout.                                 | number                         | 60                   |
| staleInterval   | Interval in seconds between checks for stale sessions.                                                                   | number                         | 60                   |
| stalePeriod     | Number of seconds that a stopped session is considered stale and will be removed from the list of sessions.              | number                         | 3600                 |
| statsInterval   | Interval in seconds for collecting system statistics                                                                     | number                         | 60                   |
| statsPrometheus | The port to serve Prometheus compatible metrics on. Set to `0` to turn off Prometheus exporting.                         | number                         | 9464                 |
| logsSentry      | The Sentry DSN (Data Source Name) to record warning and error log entries. Set to `null` to not send logs to Sentry.     | string                         | "https://7a4f9c9b0ef5474596c0066bb364e615@sentry.io/1818023" |
| peerSwarm       | The name of the peer swarm to join.                                                                                      | string, null                   | "sparkla"            |

<!-- OPTIONS-END -->
<!-- prettier-ignore-end -->

All options can be set, in decending order of priority, by:

- a command line argument e.g. `--port 80`
- an environment variable prefixed with `SPARKLA_` e.g. `SPARKLA_PORT=80`
- a `.json` or `.ini` configuration file, set using the `--config` option, or `.sparklarc` by default

## Development

```bash
npm run serve:dev
```

### Firecracker

To check for running `FirecrackerSession`s:

```bash
ps aux | firecracker
```

You can also check the logs and metrics of a Firecracker VM while it is running:

```bash
cat /tmp/vm-0d0b4b2fa73c5d4c10ed72ccac97b798530ba1b7ab8461fd313d320fb5d3562e/log.fifo
cat /tmp/vm-0d0b4b2fa73c5d4c10ed72ccac97b798530ba1b7ab8461fd313d320fb5d3562e/metrics.fifo
```

### Docker

To check for running `DockerSession`s:

```bash
docker ps
```


CHANGELOG.md
# [0.13.0](https://github.com/stencila/sparkla/compare/v0.12.3...v0.13.0) (2019-11-20)


### Features

* **DockerSession:** Restart a session if it exits prematurely. ([d218819](https://github.com/stencila/sparkla/commit/d2188195f06f5b7a10ba23a738af737bed948948))
* **Stats:** Add a session restarts metric ([d064fa4](https://github.com/stencila/sparkla/commit/d064fa4dd2e1b9ffcb3101866274ad0dcb849e91))

## [0.12.3](https://github.com/stencila/sparkla/compare/v0.12.2...v0.12.3) (2019-11-18)


### Bug Fixes

* **Stats:** Add disk usage metric ([3808995](https://github.com/stencila/sparkla/commit/380899511756f801d54311fec160462b44279df6))

## [0.12.2](https://github.com/stencila/sparkla/compare/v0.12.1...v0.12.2) (2019-11-17)


### Bug Fixes

* **Dependencies:** Upgrade Executa to 0.14.1 ([16b6884](https://github.com/stencila/sparkla/commit/16b6884430fb3a039b8de55d4293d87bde4249c8)), closes [#35](https://github.com/stencila/sparkla/issues/35)
* **Logging:** Fix how errors are sent to Sentry ([a51db1b](https://github.com/stencila/sparkla/commit/a51db1b09fe5b23d6c0a8bcd8354751593d04561)), closes [#34](https://github.com/stencila/sparkla/issues/34)

## [0.12.1](https://github.com/stencila/sparkla/compare/v0.12.0...v0.12.1) (2019-11-14)


### Bug Fixes

* **Logs:** Use log tags (aka labels)  for more useful metrics ([7292101](https://github.com/stencila/sparkla/commit/7292101ead273cedc6997ae4550e8ca20ed43415))
* **Sats:** Continuously record version ([d7068b2](https://github.com/stencila/sparkla/commit/d7068b28bcdc498c846b129aea360f233598e7e9))

# [0.12.0](https://github.com/stencila/sparkla/compare/v0.11.1...v0.12.0) (2019-11-13)


### Bug Fixes

* **Admin:** Improve logging of errors ([a0b2e85](https://github.com/stencila/sparkla/commit/a0b2e85bed18f9c12d1a6c0ab81cb524018ea0cc))
* **DockerSession:** Avoid error re CPU quota less than 1000 ([44569ba](https://github.com/stencila/sparkla/commit/44569ba13d6427c3bf4ef7afcc21e65af7c73ddc))


### Features

* **Logging:** Send logs to Sentry and record in statistics ([3b88e63](https://github.com/stencila/sparkla/commit/3b88e63219394ceeadb5454760884561d85b47d2))

## [0.11.1](https://github.com/stencila/sparkla/compare/v0.11.0...v0.11.1) (2019-11-13)


### Bug Fixes

* **ManagerServer:** Remove redundant stop method ([d014b6c](https://github.com/stencila/sparkla/commit/d014b6c75b225a3f8fa960a03cdb6627ed9c9583))
* **Package:** Include static files in dist ([ea93919](https://github.com/stencila/sparkla/commit/ea9391925b9d368bebb05b599fdc792393fedf90))

# [0.11.0](https://github.com/stencila/sparkla/compare/v0.10.0...v0.11.0) (2019-11-13)


### Bug Fixes

* **Packaging:** Change how package.json is read ([a3d514b](https://github.com/stencila/sparkla/commit/a3d514bd203115da6bea250a51022adaa4190679))


### Features

* **Stats:** Add package version number to stats ([ddb3dd6](https://github.com/stencila/sparkla/commit/ddb3dd6c90daa25914aa8228ed60ba784745b207))

# [0.10.0](https://github.com/stencila/sparkla/compare/v0.9.0...v0.10.0) (2019-11-12)


### Bug Fixes

* **Admin:** Fix JWT token query param ([0ce9648](https://github.com/stencila/sparkla/commit/0ce964868f41896f80abd18e90943af4cec56522))
* **CLI:** Add help command ([576557b](https://github.com/stencila/sparkla/commit/576557b283e4477bc30c2d49a2b82567558db601))
* **Package:** Update Executa ([74917f9](https://github.com/stencila/sparkla/commit/74917f9e88b757f53862b12b30743fbe5741f4e4))


### Features

* **admin page:** Add peer list to admin page ([e4359f2](https://github.com/stencila/sparkla/commit/e4359f268b709c0c55b5f4dd5fa8163b338b903f))
* **Manager:**  Pass requests through to other instances ([a0d3787](https://github.com/stencila/sparkla/commit/a0d3787d6cbe21810fe99534651fa1ba3826a702))
* **Manager:** Keep track of allocatable resources ([48ae2e9](https://github.com/stencila/sparkla/commit/48ae2e9cc71bd02aac831503d9021f1aeee01af7))
* **Peers:** Add discovery of peers ([53ef8b3](https://github.com/stencila/sparkla/commit/53ef8b3e797a421f182c207cb37b29f364f2020a))

# [0.9.0](https://github.com/stencila/sparkla/compare/v0.8.0...v0.9.0) (2019-11-11)


### Features

* **Config:** Upgrade Executa and Logga and add JWT secret to config ([fa7cd61](https://github.com/stencila/sparkla/commit/fa7cd61d829e9a6d017ce6a24bc0fed3724bcee6))

# [0.8.0](https://github.com/stencila/sparkla/compare/v0.7.0...v0.8.0) (2019-11-11)


### Features

* **Configuration:** Configure from files, env vars or argv ([96757b0](https://github.com/stencila/sparkla/commit/96757b0fc91f4332ffacff782bf399cbbf5fef0c))
* **Routable session ids:** Change session ids so that they can be routed ([eb4150b](https://github.com/stencila/sparkla/commit/eb4150b96fa2ee6b8bfa877552b4a180261af470))

# [0.7.0](https://github.com/stencila/sparkla/compare/v0.6.0...v0.7.0) (2019-11-07)


### Bug Fixes

* **Admin page:** Reduce noise in display of session instances ([99836ea](https://github.com/stencila/sparkla/commit/99836eabc33beabfc96756f206708d7d77e7458b))
* **Dependencies:** Update Executa version ([ce16a2a](https://github.com/stencila/sparkla/commit/ce16a2a3c5a5d324cf9848553b86f68b6bea5d9f))
* **Dependencies:** Upgrade Executa version ([7913fa9](https://github.com/stencila/sparkla/commit/7913fa9e6647c1451ce1684fcac52de197ad315a))
* **Dependencies:** Upgrade Executa version ([c756937](https://github.com/stencila/sparkla/commit/c756937ba8d25fa0c08ac140065a419b4c2459df))
* **DockerSession:** Adapt to new Schema version; other fixes ([eb8e7a8](https://github.com/stencila/sparkla/commit/eb8e7a8e485f51eb656bf7664b1546aad95272f5))
* **Manager:** Fix ending of sessions when client disconnects ([0e95ef6](https://github.com/stencila/sparkla/commit/0e95ef69ade250d499b7ef7064776625ae1a55ad))
* **Manager:** Handle missing session id better ([d23c7b4](https://github.com/stencila/sparkla/commit/d23c7b4bc2bc65d08a28bf75fcd5b50db3b78f52))

host/Config.ts
export class Config {
  /**
   * Output debug level log entries?
   */
  debug = false

  /**
   * The host address that the server should listen on.
   */
  host = '127.0.0.1'

  /**
   * The port that the server should listen on.
   */
  port = 9000

  /**
   * The JWT secret to use to sign and verify JWT tokens.
   * If `null` then a random secret will be generated.
   */
  jwtSecret: string | null = null

  /**
   * The class of sessions created.
   */
  sessionType: 'firecracker' | 'docker' = 'docker'

  /**
   * The total number of CPUs that can be allocated to sessions.
   * `null` = use the number of CPUs on the machine.
   */
  cpuTotal: number | null = null

  /**
   * The total number amount of memory (Gib) that can be allocated to sessions.
   * `null` = use the total amount of memory on the machine.
   */
  memoryTotal: number | null = null

  /**
   * Interval in seconds between checks
   * for expired sessions.
   */
  expiryInterval = 15

  /**
   * Number of seconds to provide clients with
   * a warning prior to reaching maximum session duration.
   */
  durationWarning = 600

  /**
   * Number of seconds to provide clients with
   * a warning prior to a reaching session timeout.
   */
  timeoutWarning = 60

  /**
   * Interval in seconds between checks
   * for stale sessions.
   */
  staleInterval = 60

  /**
   * Number of seconds that a stopped session is
   * considered stale and will be removed from the list of sessions.
   */
  stalePeriod = 3600

  /**
   * Interval in seconds for collecting system statistics
   */
  statsInterval = 60

  /**
   * The port to serve Prometheus compatible metrics on.
   * Metrics will be exported at https://localhost:{port}/metrics.
   * Set to `0` to turn off Prometheus exporting.
   */
  statsPrometheus = 9464

  /**
   * The name of the peer swarm to join.
   */
  peerSwarm: string | null = 'sparkla'
}
